services/src/main/java/org/keycloak/protocol/oidc/grants/device/endpoints/DeviceEndpoint.java

```diff
@@ -0,0 +1,488 @@
+import org.keycloak.representations.OAuth2DeviceAuthorizationResponse;
+import org.keycloak.saml.common.util.StringUtil;
+import org.keycloak.services.CorsErrorResponseException;
+import org.keycloak.services.ErrorPageException;
+import org.keycloak.services.ServicesLogger;
+import org.keycloak.services.Urls;
+import org.keycloak.services.managers.AuthenticationManager;
+import org.keycloak.services.managers.ClientSessionCode;
+import org.keycloak.services.messages.Messages;
+import org.keycloak.services.resource.RealmResourceProvider;
+import org.keycloak.services.resources.Cors;
+import org.keycloak.services.resources.SessionCodeChecks;
+import org.keycloak.services.util.CacheControlUtil;
+import org.keycloak.sessions.AuthenticationSessionModel;
+import org.keycloak.util.JsonSerialization;
+import org.keycloak.util.TokenUtil;
+
+import javax.ws.rs.Consumes;
+import javax.ws.rs.GET;
+import javax.ws.rs.HttpMethod;
+import javax.ws.rs.POST;
+import javax.ws.rs.Path;
+import javax.ws.rs.Produces;
+import javax.ws.rs.QueryParam;
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.MultivaluedMap;
+import javax.ws.rs.core.Response;
+
+import static org.keycloak.protocol.oidc.endpoints.AuthorizationEndpoint.LOGIN_SESSION_NOTE_ADDITIONAL_REQ_PARAMS_PREFIX;
+import static org.keycloak.protocol.oidc.grants.device.DeviceGrantType.OAUTH2_DEVICE_USER_CODE;
+import static org.keycloak.protocol.oidc.grants.device.DeviceGrantType.OAUTH2_USER_CODE_VERIFY;
+import static org.keycloak.services.resources.LoginActionsService.SESSION_CODE;
+
+/**
+ * @author <a href="mailto:h2-wada@nri.co.jp">Hiroyuki Wada</a>
+ */
+public class DeviceEndpoint extends AuthorizationEndpointBase implements RealmResourceProvider {
+
+    protected static final Logger logger = Logger.getLogger(DeviceEndpoint.class);
+
+    protected Cors cors;
+
+    public DeviceEndpoint(RealmModel realm, EventBuilder event) {
+        super(realm, event);
+    }
+
+    /**
+     * Handles device authorization requests.
+     *
+     * @return the device authorization response.
+     */
+    @Path("/auth")
+    @POST
+    @Consumes(MediaType.APPLICATION_FORM_URLENCODED)
+    @Produces(MediaType.APPLICATION_JSON)
+    public Response handleDeviceRequest() {
+        logger.trace("Processing @POST request");
+        cors = Cors.add(httpRequest).auth().allowedMethods("POST").auth().exposedHeaders(Cors.ACCESS_CONTROL_ALLOW_METHODS);
+        event.event(EventType.OAUTH2_DEVICE_AUTH);
+
+        checkSsl();
+        checkRealm();
+
+        ClientModel client = authenticateClient();
+
+        AuthorizationEndpointRequest request = AuthorizationEndpointRequestParserProcessor.parseRequest(event, session, client,
+            httpRequest.getDecodedFormParameters());
+
+        if (!TokenUtil.isOIDCRequest(request.getScope())) {
+            ServicesLogger.LOGGER.oidcScopeMissing();
+        }
+
+        createAuthenticationSession(client, request);
```

Comment:
I am not sure what exactly is the reason of create authenticationSession here. Is it really needed? It looks that endpoint to deviceUrl needs just to create deviceCode and userCode and save the stuff to infinispan actionTokens cache. But authenticationSession is probably not needed? I think it is sufficient to create it later once user opens the URL of the browser?

Reply:
I agree, not needed.

Reply:
I fixed.